### DISCUSSION TITLE
How to add dummy bytes to make the line width multiple of 64/128 bytes?

### DISCUSSION THREAD

**Discussion by markszente** [slint-ui/slint#5406] [Q&A] [answered]
Hi guys,

I'm using `SoftwareRenderer` in a `no_std` environment. ST's [AN4861](https://www.st.com/resource/en/application_note/an4861-introduction-to-lcdtft-display-controller-ltdc-on-stm32-mcus-stmicroelectronics.pdf) speaks in detail about how we should 

> add a number of dummy bytes at the end of every line of pixels to match the closest frame line width
> multiple of 64/128 bytes (for AHB/AXI)

for optimal performance. 

My question is, how would I do it with Slint? I'm using 24bpp (RGB888) color, and the width of the screen is 800. That's 2,400 bytes if my math is correct, not a multiple of 64/128. I would need to add 32 bytes to the end of each line (and then set the pitch in LTDC, but that's beyond the point). Is it possible? 

Thanks in advance.

**Comment by ogoffart** [comment, marked as answer]
If you are using the SoftwareRenderer::render, you can allocate a bigger buffer and change the stride. But unfortunately, the stride being in pixel and not byte, you can't add 32 bytes :-(. 

What you can do is add 12 pixels (288 bytes) at the end of each lines. That's some wasted spaces, but maybe that helps.

In a future version of Slint we should change the software renderer to have a `trait TargetBuffer` instead of `TargetPixel` which would have a function to retrieve lines.

Another solution is to render line by line into the buffer using the `render_by_line` function
